presto-main/src/main/java/com/facebook/presto/operator/SpatialLookupJoinOperator.java

```diff
@@ -0,0 +1,259 @@
+
+        private boolean closed;
+
+        public SpatialLookupJoinOperatorFactory(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<Type> probeTypes,
+                List<Integer> probeOutputChannels,
+                int probeGeometryChannel,
+                PagesSpatialIndexFactory pagesSpatialIndexFactory,
+                BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = planNodeId;
+            this.probeTypes = ImmutableList.copyOf(probeTypes);
+            this.probeOutputTypes = probeOutputChannels.stream()
+                    .map(probeTypes::get)
+                    .collect(toImmutableList());
+            this.buildOutputTypes = pagesSpatialIndexFactory.getOutputTypes();
+            this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+            this.probeGeometryChannel = probeGeometryChannel;
+            this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+            this.spatialRelationshipTest = spatialRelationshipTest;
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return ImmutableList.<Type>builder()
+                    .addAll(probeOutputTypes)
+                    .addAll(buildOutputTypes)
+                    .build();
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, HashBuilderOperator.class.getSimpleName());
+            return new SpatialLookupJoinOperator(operatorContext, getTypes(), probeTypes, probeOutputChannels, probeGeometryChannel, pagesSpatialIndexFactory.createPagesSpatialIndex(), spatialRelationshipTest);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            this.closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new SpatialLookupJoinOperatorFactory(operatorId, planNodeId, probeTypes, probeOutputChannels, probeGeometryChannel, pagesSpatialIndexFactory, spatialRelationshipTest);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Integer> probeOutputChannels;
+    private final int probeGeometryChannel;
+    private final ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final PageBuilder pageBuilder;
+    private Page probe;
+    private int probePosition;
+    private List<Long> joinAddresses;
+    private int joinAddressIndex;
+    private Page outputPage;
+    private boolean finishing;
+    private boolean finished;
+
+    public SpatialLookupJoinOperator(
```

Comment:
I don't think we need `Lookup` in any of these class names... For historical context, we switched from `Hash` the 'Lookup` in the main join operator when we added support for non-hash data structures (e.g., index) for the join, but here I feel like it doesn't help.